File: swift/Sources/OpenID4VP/Dcql.swift
import Foundation
import SdJwt

public struct DcqlError: Error, CustomStringConvertible {
    public let description: String
    init(_ description: String) { self.description = description }
}

/// A single element of a DCQL claims path (OpenID4VP §6.4): key, array index, or wildcard (JSON null).
public enum PathElement: Equatable {
    case key(String)
    case index(Int)
    case wildcard
}

public struct ClaimQuery {
    public let id: String?
    public let path: [PathElement]
    public let values: [JsonValue]?
}

public struct CredentialMeta {
    public let vctValues: [String]?
    public let doctypeValue: String?
}

public struct CredentialQuery {
    public let id: String
    public let format: String
    public let meta: CredentialMeta?
    public let claims: [ClaimQuery]
    public let claimSets: [[String]]?
}

public struct CredentialSetQuery {
    public let options: [[String]]
    public let required: Bool
}

public struct DcqlQuery {
    public let credentials: [CredentialQuery]
    public let credentialSets: [CredentialSetQuery]?

    public static func parse(_ obj: JsonValue) throws -> DcqlQuery {
        guard case let .arr(credArr)? = obj["credentials"] else { throw DcqlError("dcql: missing 'credentials'") }
        let creds = try credArr.map { try parseCredentialQuery($0) }
        guard !creds.isEmpty else { throw DcqlError("dcql: 'credentials' is empty") }
        var sets: [CredentialSetQuery]?
        if case let .arr(setArr)? = obj["credential_sets"] {
            sets = try setArr.map { try parseCredentialSet($0) }
        }
        return DcqlQuery(credentials: creds, credentialSets: sets)
    }

    private static func parseCredentialQuery(_ o: JsonValue) throws -> CredentialQuery {
        guard case let .str(id)? = o["id"] else { throw DcqlError("credential query: missing 'id'") }
        guard case let .str(format)? = o["format"] else { throw DcqlError("credential query '\(id)': missing 'format'") }
        var meta: CredentialMeta?
        if let m = o["meta"] {
            var vct: [String]?
            if case let .arr(items)? = m["vct_values"] {
                vct = items.compactMap { if case let .str(s) = $0 { return s } else { return nil } }
            }
            var doctype: String?
            if case let .str(d)? = m["doctype_value"] { doctype = d }
            meta = CredentialMeta(vctValues: vct, doctypeValue: doctype)
        }
        var claims: [ClaimQuery] = []
        if case let .arr(cArr)? = o["claims"] { claims = try cArr.map { try parseClaimQuery($0) } }
        var claimSets: [[String]]?
        if case let .arr(csArr)? = o["claim_sets"] {
            claimSets = try csArr.map { set in
                guard case let .arr(ids) = set else { throw DcqlError("claim_sets must be arrays") }
                return try ids.map { if case let .str(s) = $0 { return s } else { throw DcqlError("claim_sets entries must be strings") } }
            }
        }
        // mdoc (ISO 18013-5): a claim path addresses [namespace, data element], both strings.
        // Require the first two segments to be string keys (>=2, not strictly ==2 — a deeper
        // path may index into a structured element value).
        if format == "mso_mdoc" {
            for c in claims {
                let firstTwoAreKeys = c.path.count >= 2 && {
                    if case .key = c.path[0], case .key = c.path[1] { return true } else { return false }
                }()
                if !firstTwoAreKeys {
                    throw DcqlError("credential query '\(id)': mso_mdoc claim path must start with [namespace, element] (two strings)")
                }
            }
        }
        return CredentialQuery(id: id, format: format, meta: meta, claims: claims, claimSets: claimSets)
    }

    private static func parseClaimQuery(_ o: JsonValue) throws -> ClaimQuery {
        guard case let .arr(pathArr)? = o["path"] else { throw DcqlError("claim: missing 'path'") }
        let path = try pathArr.map { el -> PathElement in
            switch el {
            case let .str(s): return .key(s)
            case let .numInt(i): return .index(Int(i))
            case .null: return .wildcard
            default: throw DcqlError("path element must be string, int, or null")
            }
        }
        var values: [JsonValue]?
        if case let .arr(v)? = o["values"] { values = v }
        var id: String?
        if case let .str(s)? = o["id"] { id = s }
        return ClaimQuery(id: id, path: path, values: values)
    }

    private static func parseCredentialSet(_ o: JsonValue) throws -> CredentialSetQuery {
        guard case let .arr(optArr)? = o["options"] else { throw DcqlError("credential_set: missing 'options'") }
        let options = try optArr.map { opt -> [String] in
            guard case let .arr(ids) = opt else { throw DcqlError("options must be arrays") }
            return try ids.map { if case let .str(s) = $0 { return s } else { throw DcqlError("option ids must be strings") } }
        }
        var required = true
        if case let .bool(b)? = o["required"] { required = b }
        return CredentialSetQuery(options: options, required: required)
    }
}

File: kotlin/openid4vp/src/main/kotlin/com/hopae/eudi/wallet/vp/Dcql.kt
package com.hopae.eudi.wallet.vp

import com.hopae.eudi.wallet.sdjwt.JsonValue

class DcqlException(message: String) : Exception(message)

/** A single element of a DCQL claims path (OpenID4VP §6.4): key, array index, or wildcard (JSON null). */
sealed interface PathElement {
    data class Key(val name: String) : PathElement
    data class Index(val index: Int) : PathElement

    /** JSON `null`: selects every element of an array (the null array-wildcard). */
    data object Wildcard : PathElement
}

data class ClaimQuery(
    val id: String?,
    val path: List<PathElement>,
    /** If present, the claim matches only when its value is one of these. */
    val values: List<JsonValue>?,
)

/** DCQL meta constraints — vct for SD-JWT VC, doctype for mdoc. */
data class CredentialMeta(
    val vctValues: List<String>?,
    val doctypeValue: String?,
)

data class CredentialQuery(
    val id: String,
    val format: String,
    val meta: CredentialMeta?,
    val claims: List<ClaimQuery>,
    /** Alternative sets of claim ids; at least one set must be fully satisfiable. */
    val claimSets: List<List<String>>?,
    /** §6.1: whether more than one Credential may be returned for this query. Default false (exactly one). */
    val multiple: Boolean = false,
    /**
     * §6.1: whether the Verifier requires a Cryptographic Holder Binding proof. Default true. When false the
     * Verifier accepts a Credential without holder binding, so an SD-JWT VC may be presented with no KB-JWT.
     */
    val requireCryptographicHolderBinding: Boolean = true,
)

data class CredentialSetQuery(
    val options: List<List<String>>,
    val required: Boolean,
)

data class DcqlQuery(
    val credentials: List<CredentialQuery>,
    val credentialSets: List<CredentialSetQuery>?,
) {
    companion object {
        fun parse(obj: JsonValue.Obj): DcqlQuery {
            val creds = (obj["credentials"] as? JsonValue.Arr)?.items?.map { parseCredentialQuery(it.asObj("credential")) }
                ?: throw DcqlException("dcql: missing 'credentials'")
            if (creds.isEmpty()) throw DcqlException("dcql: 'credentials' is empty")
            val sets = (obj["credential_sets"] as? JsonValue.Arr)?.items?.map { parseCredentialSet(it.asObj("credential_set")) }
            return DcqlQuery(creds, sets)
        }

        private fun parseCredentialQuery(o: JsonValue.Obj): CredentialQuery {
            val id = o.str("id") ?: throw DcqlException("credential query: missing 'id'")
            val format = o.str("format") ?: throw DcqlException("credential query '$id': missing 'format'")
            val metaObj = o["meta"] as? JsonValue.Obj
            val meta = metaObj?.let {
                CredentialMeta(
                    vctValues = (it["vct_values"] as? JsonValue.Arr)?.items?.mapNotNull { v -> (v as? JsonValue.Str)?.value },
                    doctypeValue = (it["doctype_value"] as? JsonValue.Str)?.value,
                )
            }
            val claims = (o["claims"] as? JsonValue.Arr)?.items?.map { parseClaimQuery(it.asObj("claim")) } ?: emptyList()
            val claimSets = (o["claim_sets"] as? JsonValue.Arr)?.items?.map { set ->
                (set as? JsonValue.Arr)?.items?.map { id2 -> (id2 as? JsonValue.Str)?.value ?: throw DcqlException("claim_sets entries must be strings") }
                    ?: throw DcqlException("claim_sets must be arrays")
            }
            // mdoc (ISO 18013-5): a claim path addresses [namespace, data element], both strings.
            // Require the first two segments to be string keys (>=2, not strictly ==2 — a deeper
            // path may index into a structured element value).
            if (format == "mso_mdoc") {
                claims.forEach { c ->
                    if (c.path.size < 2 || c.path[0] !is PathElement.Key || c.path[1] !is PathElement.Key) {
                        throw DcqlException("credential query '$id': mso_mdoc claim path must start with [namespace, element] (two strings)")
                    }
                }
            }
            val multiple = (o["multiple"] as? JsonValue.Bool)?.value ?: false
            val requireBinding = (o["require_cryptographic_holder_binding"] as? JsonValue.Bool)?.value ?: true
            return CredentialQuery(id, format, meta, claims, claimSets, multiple, requireBinding)
        }

        private fun parseClaimQuery(o: JsonValue.Obj): ClaimQuery {
            val pathArr = o["path"] as? JsonValue.Arr ?: throw DcqlException("claim: missing 'path'")
            val path = pathArr.items.map { el ->
                when (el) {
                    is JsonValue.Str -> PathElement.Key(el.value)
                    is JsonValue.NumInt -> PathElement.Index(el.value.toInt())
                    JsonValue.Null -> PathElement.Wildcard
                    else -> throw DcqlException("path element must be string, int, or null")
                }
            }
            val values = (o["values"] as? JsonValue.Arr)?.items
            return ClaimQuery(o.str("id"), path, values)
        }

        private fun parseCredentialSet(o: JsonValue.Obj): CredentialSetQuery {
            val options = (o["options"] as? JsonValue.Arr)?.items?.map { opt ->
                (opt as? JsonValue.Arr)?.items?.map { (it as? JsonValue.Str)?.value ?: throw DcqlException("option ids must be strings") }
                    ?: throw DcqlException("options must be arrays")
            } ?: throw DcqlException("credential_set: missing 'options'")
            val required = (o["required"] as? JsonValue.Bool)?.value ?: true
            return CredentialSetQuery(options, required)
        }

        private fun JsonValue.asObj(what: String): JsonValue.Obj =
            this as? JsonValue.Obj ?: throw DcqlException("$what must be an object")

        private fun JsonValue.Obj.str(name: String): String? = (this[name] as? JsonValue.Str)?.value
    }
}
